aspnetcore/blazor/security/blazor-web-app-with-entra.md

```diff
@@ -0,0 +1,172 @@
+
+Use the [Secret Manager tool](xref:security/app-secrets) to store the server app's client secret under the configuration key `AzureAd:ClientSecret`.
+
+Create a client secret in the app's Entra ID registration in the Entra or Azure portal (**Manage** > **Certificates & secrets** > **New client secret**). Use the **Value** of the new secret in the following guidance.
+
+Execute the following command in a command shell from the server project's directory, such as the Developer PowerShell command shell in Visual Studio. The `{SECRET}` placeholder is the client secret obtained from the app's registration:
+
+```dotnetcli
+dotnet user-secrets set "AzureAd:ClientSecret" "{SECRET}"
+```
+
+If using Visual Studio, you can confirm the secret is set by right-clicking the server project in **Solution Explorer** and selecting **Manage User Secrets**.
+
+### Configure the app
+
+In the server project's app settings file (`appsettings.json`), provide the app's `AzureAd` section configuration. Obtain the application (client) ID, tenant (publisher) domain, and directory (tenant) ID from the app's registration in the Entra or Azure portal:
+
+```json
+"AzureAd": {
+  "CallbackPath": "/signin-oidc",
+  "ClientId": "{CLIENT ID}",
+  "Domain": "{DOMAIN}",
+  "Instance": "https://login.microsoftonline.com/",
+  "ResponseType": "code",
+  "TenantId": "{TENANT ID}"
+},
+```
+
+Placeholders in the preceding example:
+
+* `{CLIENT ID}`: The application (client) ID.
+* `{DOMAIN}`: The tenant (publisher) domain.
+* `{TENANT ID}`: The directory (tenant) ID.
+
+Example:
+
+```json
+"AzureAd": {
+  "CallbackPath": "/signin-oidc",
+  "ClientId": "41451fa7-82d9-4673-8fa5-69eff5a761fd",
+  "Domain": "contoso.onmicrosoft.com",
+  "Instance": "https://login.microsoftonline.com/",
+  "ResponseType": "code",
+  "TenantId": "e86c78e2-8bb4-4c41-aefd-918e0565a45e"
+},
+```
+
+The callback path (`CallbackPath`) must match the redirect URI (login callback path) configured when registering the application in the Entra or Azure portal. Paths are configured in the **Authentication** blade of the app's registration. The default value of `CallbackPath` is `/signin-oidc` for a registered redirect URI of `https://localhost/signin-oidc` (a port isn't required).
+
+<!-- UPDATE 9.0 Replace the following with the INCLUDE link after 
+                https://github.com/dotnet/AspNetCore.Docs/pull/33749
+                merges. 
+                
+                [!INCLUDE[](~/blazor/security/includes/secure-authentication-flows.md)]
+                -->
+
+> [!WARNING]
+> Don't store app secrets, connection strings, credentials, passwords, personal identification numbers (PINs), private C#/.NET code, or private keys/tokens in client-side code, which is ***always insecure***. In test/staging and production environments, server-side Blazor code and web APIs should use secure authentication flows that avoid maintaining credentials within project code, configuration files, and environment variables.
```

Comment:
```suggestion
> Don't store app secrets, connection strings, credentials, passwords, personal identification numbers (PINs), private C#/.NET code, or private keys/tokens in client-side code, which is ***always insecure***. In test/staging and production environments, server-side Blazor code and web APIs should use secure authentication flows that avoid maintaining credentials within project code or configuration files.
```

While environment variables have downsides compared to Azure Key Vault, we still recommend environment variables for configuring secrets over the alternatives like project code and configuration files.

> Secrets shouldn't be deployed with the app. Instead, production secrets should be accessed through a controlled means like environment variables or Azure Key Vault.

https://learn.microsoft.com/aspnet/core/security/app-secrets

Reply:
There was a security initiative by docs management on this that I don't have full clarity on.

@Rick-Anderson ... Did the communication that the docs team received from docs management say that env vars can be used in production? My limited understanding was that they are now saying that they don't want to pitch env vars for production.

Separate question was on if env vars are ok for test/staging environments. I was waiting on @danroth27 to make a ruling on it (or Barry or whoever makes such decisions).

Reply:
@blowdart Do you have opinions about environment variables for secrets?

I see the downsides. Dumping environment variables is one of the first things an attacker will try if they are able to do so. However, it's still a lot better than putting secrets somewhere it might get checked into source control. If we say that user-secrets and Azure Key Vault are the only viable options, people might just ignore our advice as mere upselling.

I'm not saying that we should tout environment variables as the best possible option for configuring secrets. We already [warn people about the downsides](https://learn.microsoft.com/aspnet/core/security/app-secrets?view=aspnetcore-8.0#environment-variables), but I don't think we should warn against it as strongly as we warn against putting secrets in project code or config files.

Reply:
See [Secure authentication flows](https://learn.microsoft.com/en-us/aspnet/core/security/?view=aspnetcore-8.0#additional-resources)

*When the app is deployed to a test server, an environment variable can be used to set the connection string to a test database server. For more information, see [Configuration](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-8.0). An environment variable should NEVER be used to store a production connection string.*

@halter73 that was approved by @javiercn and steve. That should be enough for them to connect the dots, env's are the next safest approach..

Reply:
I kind of agree with Stephen here. Saying "Use keyvault" in production isn't that realistic, and assumes everyone is just running on Azure. Sure, env variables are a last resort but frankly it's how lots of things work. You could at least wave to other cloud platforms having their own solutions?

Reply:
Thanks everyone. I'll make updates to this PR in the morning, and I'll adjust the other PR to match.

@Rick-Anderson ... I'll wait for you to look at the changes on the other PR before merging it. I'll ping u when it's ready.